<h1><center>Quartic DataSet Solution </center></h1>

In [1]:
# Importing basic libraries

import pandas as pd
import numpy as np
import random as rnd

# visualization
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

Read the dataset Train and Test.

In [2]:
# Training DataSet with Target field
data_train=pd.read_csv('E:/softwares/Ubuntu/dataset/quartic/ds_data/data_train.csv')
data_train.head()

id  num1  num2  num3  num4  num5  num6  num7  num8  num9   ...    cat6  \
0   0     2     5     0     1     0     0     0     0     0   ...     NaN   
1   1     1     7     0     0     1     0     0     0     0   ...     NaN   
2   2     5     9     0     0     1     0     0     0     0   ...     NaN   
3   3     0     2     1     0     0     0     0     0     0   ...     0.0   
4   4     0     0     1     0     0     0     0     0     0   ...     NaN   

   cat7  cat8  cat9  cat10  cat11  cat12  cat13  cat14  target  
0     0   1.0     4    1.0      0    0.0      1     12       0  
1     0   NaN    11    1.0      1    2.0      1     19       0  
2     0   NaN    14    1.0      1    2.0      1     60       0  
3     0   1.0    11    1.0      1    3.0      1    104       0  
4     0   NaN    14    1.0      1    2.0      1     82       0  

[5 rows x 58 columns]

In [3]:
# Test DataSet without Target Field
data_test=pd.read_csv('E:/softwares/Ubuntu/dataset/quartic/ds_data/data_test.csv')
data_test.head()

id  num1  num2  num3  num4  num5  num6  num7  num8  num9  ...    cat5  \
0   0     0     8     0     1     0     0     0     0     0  ...     1.0   
1   1     4     5     0     0     0     1     0     0     0  ...     1.0   
2   2     5     3     0     0     0     1     0     0     0  ...     1.0   
3   3     0     6     1     0     0     0     0     0     0  ...     1.0   
4   4     5     7     0     0     0     1     0     0     0  ...     1.0   

   cat6  cat7  cat8  cat9  cat10  cat11  cat12  cat13  cat14  
0   NaN     0   NaN     1    1.0      1    2.0      1     65  
1   NaN     0   0.0    11    1.0      1    0.0      1    103  
2   NaN     0   NaN    14    1.0      1    2.0      1     29  
3   NaN     0   NaN     1    1.0      1    2.0      1     40  
4   NaN     0   NaN    11    1.0      1    2.0      1    101  

[5 rows x 57 columns]

### Basic Insight of Data

After reading the dataset in dataframe ,we are going to explore the dataset to understand it.

In [17]:
# Getting the info about data its datatyes number of rows and the name of column field.
data_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 596000 entries, 0 to 595999
Data columns (total 58 columns):
id        596000 non-null int64
num1      596000 non-null int64
num2      596000 non-null int64
num3      596000 non-null int64
num4      596000 non-null int64
num5      596000 non-null int64
num6      596000 non-null int64
num7      596000 non-null int64
num8      596000 non-null int64
num9      596000 non-null int64
num10     596000 non-null int64
num11     596000 non-null int64
num12     596000 non-null int64
num13     596000 non-null int64
num14     596000 non-null int64
num15     596000 non-null int64
num16     596000 non-null float64
num17     596000 non-null float64
num18     488091 non-null float64
num19     595995 non-null float64
num20     595999 non-null float64
num21     596000 non-null float64
num22     553333 non-null float64
num23     596000 non-null float64
der1      596000 non-null float64
der2      596000 non-null float64
der3      596000 non-null float64
der

In [18]:
data_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 892816 entries, 0 to 892815
Data columns (total 57 columns):
id       892816 non-null int64
num1     892816 non-null int64
num2     892816 non-null int64
num3     892816 non-null int64
num4     892816 non-null int64
num5     892816 non-null int64
num6     892816 non-null int64
num7     892816 non-null int64
num8     892816 non-null int64
num9     892816 non-null int64
num10    892816 non-null int64
num11    892816 non-null int64
num12    892816 non-null int64
num13    892816 non-null int64
num14    892816 non-null int64
num15    892816 non-null int64
num16    892816 non-null float64
num17    892816 non-null float64
num18    731132 non-null float64
num19    892815 non-null float64
num20    892816 non-null float64
num21    892816 non-null float64
num22    829011 non-null float64
num23    892816 non-null float64
der1     892816 non-null float64
der2     892816 non-null float64
der3     892816 non-null float64
der4     892816 non-null int64

By looking in the info we can have following conclusions of the DataSet.

1. The datatypes of the variables field looks good as per the data content in it for both *data_train* and *data_test* , except few    int defined as float but we can ignore that.
2. There are certain amount of missing values present in both *data_train* and *data_test* which needs to identified and dealt with.

We will continue our analysis on this.

In [19]:
#Understanding the statistical summary of each column.
data_train.describe()

id           num1           num2           num3  \
count  596000.000000  596000.000000  596000.000000  596000.000000   
mean   297999.500000       1.900423       4.423356       0.393743   
std    172050.524555       1.983855       2.699861       0.488579   
min         0.000000       0.000000       0.000000       0.000000   
25%    148999.750000       0.000000       2.000000       0.000000   
50%    297999.500000       1.000000       4.000000       0.000000   
75%    446999.250000       3.000000       6.000000       1.000000   
max    595999.000000       7.000000      11.000000       1.000000   

                num4           num5           num6           num7  \
count  596000.000000  596000.000000  596000.000000  596000.000000   
mean        0.257015       0.163933       0.185309       0.000374   
std         0.436988       0.370215       0.388548       0.019340   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         1.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

                num8           num9      ...                 cat6  \
count  596000.000000  596000.000000      ...        184208.000000   
mean        0.001693       0.009435      ...             0.601744   
std         0.041111       0.096672      ...             0.489540   
min         0.000000       0.000000      ...             0.000000   
25%         0.000000       0.000000      ...             0.000000   
50%         0.000000       0.000000      ...             1.000000   
75%         0.000000       0.000000      ...             1.000000   
max         1.000000       1.000000      ...             1.000000   

                cat7           cat8           cat9          cat10  \
count  596000.000000  329072.000000  596000.000000  584497.000000   
mean        0.725106       0.525365       6.555000       0.947608   
std         2.153319       0.499357       5.501419       0.222817   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       1.000000       1.000000   
50%         0.000000       1.000000       7.000000       1.000000   
75%         0.000000       1.000000      11.000000       1.000000   
max         9.000000       1.000000      17.000000       1.000000   

               cat11          cat12          cat13          cat14  \
count  596000.000000  595430.000000  596000.000000  596000.000000   
mean        0.832077       1.331161       0.992136      62.215570   
std         0.373798       0.976544       0.091613      33.012105   
min         0.000000       0.000000       0.000000       1.000000   
25%         1.000000       0.000000       1.000000      32.000000   
50%         1.000000       2.000000       1.000000      65.000000   
75%         1.000000       2.000000       1.000000      93.000000   
max         1.000000       4.000000       2.000000     104.000000   

              target  
count  596000.000000  
mean        0.036436  
std         0.187373  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 58 columns]

In [20]:
#Statistical summary for Test DataSet
data_test.describe()

id           num1           num2           num3  \
count  8.928160e+05  892816.000000  892816.000000  892816.000000   
mean   7.441535e+05       1.902371       4.413734       0.393246   
std    4.296830e+05       1.986503       2.700149       0.488471   
min    0.000000e+00       0.000000       0.000000       0.000000   
25%    3.720218e+05       0.000000       2.000000       0.000000   
50%    7.443070e+05       1.000000       4.000000       0.000000   
75%    1.116308e+06       3.000000       6.000000       1.000000   
max    1.488026e+06       7.000000      11.000000       1.000000   

                num4           num5           num6           num7  \
count  892816.000000  892816.000000  892816.000000  892816.000000   
mean        0.257191       0.163659       0.185905       0.000373   
std         0.437086       0.369966       0.389030       0.019309   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         1.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

                num8           num9      ...                 cat5  \
count  892816.000000  892816.000000      ...        892811.000000   
mean        0.001595       0.009376      ...             0.829951   
std         0.039905       0.096375      ...             0.375676   
min         0.000000       0.000000      ...             0.000000   
25%         0.000000       0.000000      ...             1.000000   
50%         0.000000       0.000000      ...             1.000000   
75%         0.000000       0.000000      ...             1.000000   
max         1.000000       1.000000      ...             1.000000   

                cat6           cat7           cat8           cat9  \
count  275905.000000  892816.000000  492457.000000  892816.000000   
mean        0.602138       0.725798       0.525717       6.564297   
std         0.489457       2.154886       0.499339       5.500485   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       1.000000   
50%         1.000000       0.000000       1.000000       7.000000   
75%         1.000000       0.000000       1.000000      11.000000   
max         1.000000       9.000000       1.000000      17.000000   

               cat10          cat11          cat12          cat13  \
count  875485.000000  892816.000000  891939.000000  892816.000000   
mean        0.947957       0.832311       1.330200       0.992134   
std         0.222114       0.373590       0.976495       0.091530   
min         0.000000       0.000000       0.000000       0.000000   
25%         1.000000       1.000000       0.000000       1.000000   
50%         1.000000       1.000000       2.000000       1.000000   
75%         1.000000       1.000000       2.000000       1.000000   
max         1.000000       1.000000       4.000000       2.000000   

               cat14  
count  892816.000000  
mean       62.284293  
std        33.003714  
min         1.000000  
25%        32.000000  
50%        65.000000  
75%        94.000000  
max       104.000000  

[8 rows x 57 columns]

Identifying and dealing with the missing values. 

In [4]:
# Count of null values for each column of data_train
data_train.isnull().sum()

id             0
num1           0
num2           0
num3           0
num4           0
num5           0
num6           0
num7           0
num8           0
num9           0
num10          0
num11          0
num12          0
num13          0
num14          0
num15          0
num16          0
num17          0
num18     107909
num19          5
num20          1
num21          0
num22      42667
num23          0
der1           0
der2           0
der3           0
der4           0
der5           0
der6           0
der7           0
der8           0
der9           0
der10          0
der11          0
der12          0
der13          0
der14          0
der15          0
der16          0
der17          0
der18          0
der19          0
cat1         217
cat2          83
cat3        5814
cat4         107
cat5           5
cat6      411792
cat7           0
cat8      266928
cat9           0
cat10      11503
cat11          0
cat12        570
cat13          0
cat14          0
target         0
dtype: int64

In [5]:
data_test.isnull().sum()

id            0
num1          0
num2          0
num3          0
num4          0
num5          0
num6          0
num7          0
num8          0
num9          0
num10         0
num11         0
num12         0
num13         0
num14         0
num15         0
num16         0
num17         0
num18    161684
num19         1
num20         0
num21         0
num22     63805
num23         0
der1          0
der2          0
der3          0
der4          0
der5          0
der6          0
der7          0
der8          0
der9          0
der10         0
der11         0
der12         0
der13         0
der14         0
der15         0
der16         0
der17         0
der18         0
der19         0
cat1        307
cat2        145
cat3       8710
cat4        160
cat5          5
cat6     616911
cat7          0
cat8     400359
cat9          0
cat10     17331
cat11         0
cat12       877
cat13         0
cat14         0
dtype: int64

#### Handling the missing values by removing it or replacing it with mean or mode.

So from above we have following columns for **data_train** with missing values:  <br>

**num18**     107909   ----> signicant number of null data , we should delete this column. Datatype is float <br>      
**num19**          5   ----> Lets replace the missing values with mode i.e '3' .<br>
   data_train['num19'].describe()<br>
   count    595995.000000  <br>
   mean          2.346109  <br> 
   std           0.832505  <br>
   min           0.000000  <br>
   25%           2.000000  <br>
   50%           3.000000  <br> 
   75%           3.000000  <br>
   max           3.000000  <br>
   Name: num19, dtype: float64 <br>
**num20**          1      ---->  1 missing value replace it with Mean i.e '0.379944'. <br>
   data_train['num20'].describe() <br>
   count    595999.000000         <br> 
   mean          0.379944         <br>
   std           0.058299         <br> 
   min           0.100000         <br>
   25%           0.316228         <br>  
   50%           0.374166         <br>
   75%           0.400000         <br>
   max           1.264911         <br>
   Name: num20, dtype: float64    <br> 
**num22**          42667   ----> significant number of missing values , better to remove this column.<br>
**cat1**              217   ----> Relace the missing value by 'Mode' i.e '1.0' here.                  <br>
**cat2**          83   ----> Relace the missing value by 'Mode' i.e '0.0' here.                  <br> 
**cat3**        5814   ----> significant number of missing values , better to remove this column. <br>    
**cat4**         107   ----> Relace the missing value by 'Mode' i.e '11.00' here.                <br>
**cat5**           5   ----> Relace the missing value by 'Mode' i.e '1.0' here.                  <br>
**cat6**      411792   ----> significant number of missing values , better to remove this column. <br>
**cat8**      266928   ----> significant number of missing values , better to remove this column. <br>
**cat10**      11503   ----> significant number of missing values , better to remove this column. <br>
**cat12**        570   ----> Relace the missing value by 'Mode' i.e '2.0' here.                   <br>
   data_train['cat12'].mode()                                                                     <br>
    0    2.0                                                                                       
    dtype: float64

Before removing the **missing values** column lets understand how it is correlated with the **target** variable.

In [22]:
data_train[['num18','num19','num20','num22','cat1','cat2','cat3','cat4','cat5','cat6','cat8','cat10','cat12','target']].corr()

num18     num19     num20     num22      cat1      cat2      cat3  \
num18   1.000000  0.052909  0.228948  0.090902  0.072256  0.094798 -0.003697   
num19   0.052909  1.000000  0.056987 -0.039720  0.019643  0.073395  0.002077   
num20   0.228948  0.056987  1.000000  0.603352 -0.110399 -0.095090 -0.019732   
num22   0.090902 -0.039720  0.603352  1.000000 -0.084583 -0.082438 -0.015164   
cat1    0.072256  0.019643 -0.110399 -0.084583  1.000000  0.146339 -0.007734   
cat2    0.094798  0.073395 -0.095090 -0.082438  0.146339  1.000000  0.012464   
cat3   -0.003697  0.002077 -0.019732 -0.015164 -0.007734  0.012464  1.000000   
cat4    0.225325  0.110833  0.126659 -0.003056 -0.000012  0.028094 -0.035321   
cat5   -0.243133 -0.090155 -0.469456 -0.264693  0.045294  0.003074  0.013975   
cat6    0.102335  0.074213 -0.025575 -0.009943  0.019917  0.055897  0.008643   
cat8    0.160010 -0.020569  0.079052  0.063671  0.022003  0.009308  0.001936   
cat10   0.002946  0.114349 -0.027071 -0.041423  0.023221  0.084247  0.004682   
cat12  -0.009745  0.119486 -0.031970 -0.223584  0.007386 -0.042835 -0.003875   
target  0.033741 -0.001249  0.038818  0.011727  0.005582  0.010176  0.032223   

            cat4      cat5      cat6      cat8     cat10     cat12    target  
num18   0.225325 -0.243133  0.102335  0.160010  0.002946 -0.009745  0.033741  
num19   0.110833 -0.090155  0.074213 -0.020569  0.114349  0.119486 -0.001249  
num20   0.126659 -0.469456 -0.025575  0.079052 -0.027071 -0.031970  0.038818  
num22  -0.003056 -0.264693 -0.009943  0.063671 -0.041423 -0.223584  0.011727  
cat1   -0.000012  0.045294  0.019917  0.022003  0.023221  0.007386  0.005582  
cat2    0.028094  0.003074  0.055897  0.009308  0.084247 -0.042835  0.010176  
cat3   -0.035321  0.013975  0.008643  0.001936  0.004682 -0.003875  0.032223  
cat4    1.000000 -0.152980  0.090812  0.039575  0.078014  0.273542  0.017353  
cat5   -0.152980  1.000000 -0.043464 -0.069384 -0.012017 -0.013924 -0.031532  
cat6    0.090812 -0.043464  1.000000 -0.059592  0.024398 -0.046967  0.020601  
cat8    0.039575 -0.069384 -0.059592  1.000000  0.024405 -0.019268  0.000830  
cat10   0.078014 -0.012017  0.024398  0.024405  1.000000  0.002326 -0.019692  
cat12   0.273542 -0.013924 -0.046967 -0.019268  0.002326  1.000000  0.005949  
target  0.017353 -0.031532  0.020601  0.000830 -0.019692  0.005949  1.000000

From above we can see that all the **missing values** attributes are having **very weak correlation** with the **target** attribute , so we can go ahead and handle the missing the values with the approach defined before.

####  Replacing column num19, cat1 , cat2 , cat4 , cat5 , cat12 by mode and  num20 by mean.

In [4]:
## DATA_TRAIN

mode1=data_train['num19'].mode()
mode2=data_train['cat1'].mode()
mode3=data_train['cat2'].mode()
mode4=data_train['cat4'].mode()
mode5=data_train['cat5'].mode()
mode6=data_train['cat12'].mode()
print(mode1 , mode2 ,mode3, mode4 , mode5 , mode6 )

## DATA_TEST

modet1=data_test['num19'].mode()
modet2=data_test['cat1'].mode()
modet3=data_test['cat2'].mode()
modet4=data_test['cat4'].mode()
modet5=data_test['cat5'].mode()
modet6=data_test['cat12'].mode()
print(modet1 , modet2 ,modet3, modet4 , modet5 , modet6 )

0    3.0
dtype: float64 0    1.0
dtype: float64 0    0.0
dtype: float64 0    11.0
dtype: float64 0    1.0
dtype: float64 0    2.0
dtype: float64
0    3.0
dtype: float64 0    1.0
dtype: float64 0    0.0
dtype: float64 0    11.0
dtype: float64 0    1.0
dtype: float64 0    2.0
dtype: float64


In [5]:
### DATA_TRAIN
# Filling the missing values with mode as calculated above: 
data_train['num19']=data_train['num19'].fillna(3)
data_train['cat1']=data_train['cat1'].fillna(1)
data_train['cat2']=data_train['cat2'].fillna(0)
data_train['cat4']=data_train['cat4'].fillna(11)
data_train['cat5']=data_train['cat5'].fillna(1)
data_train['cat12']=data_train['cat12'].fillna(2)

# Avg is getting replaced for missing value of num20 of data_train : 
avg1=data_train['num20'].astype(float).mean()
data_train['num20'].replace(np.nan,avg1,inplace=True)

### DATA_TEST

data_test['num19']=data_test['num19'].fillna(3)
data_test['cat1']=data_test['cat1'].fillna(1)
data_test['cat2']=data_test['cat2'].fillna(0)
data_test['cat4']=data_test['cat4'].fillna(11)
data_test['cat5']=data_test['cat5'].fillna(1)
data_test['cat12']=data_test['cat12'].fillna(2)

print('Done')

Done


#### Dropping the column id (as no role in determining the target) ,num18 , num22 , cat3 ,cat6 ,cat8 and cat10 as these having significant number of missing data and weak correlation.

In [6]:
## DATA_TRAIN
# simply drop whole column with significant missing values and weak correlation.
data_train_id=data_train['id']
data_train.drop(['id','num18','num22','cat3','cat6','cat8','cat10'],axis=1,inplace=True)

#DATA_TEST    , dropping the column 
data_test_id=data_test['id']
data_test.drop(['id','num18','num22','cat3','cat6','cat8','cat10'],axis=1,inplace=True)
print('Done')

Done


Now we will check whether we have any null values in our **data_train** or **data_test**. We will see that our dataset is free of null values. 

In [7]:
print(data_train.isnull().sum())
print('\n','Done')
data_test.isnull().sum()

num1      0
num2      0
num3      0
num4      0
num5      0
num6      0
num7      0
num8      0
num9      0
num10     0
num11     0
num12     0
num13     0
num14     0
num15     0
num16     0
num17     0
num19     0
num20     0
num21     0
num23     0
der1      0
der2      0
der3      0
der4      0
der5      0
der6      0
der7      0
der8      0
der9      0
der10     0
der11     0
der12     0
der13     0
der14     0
der15     0
der16     0
der17     0
der18     0
der19     0
cat1      0
cat2      0
cat4      0
cat5      0
cat7      0
cat9      0
cat11     0
cat12     0
cat13     0
cat14     0
target    0
dtype: int64

 Done


num1     0
num2     0
num3     0
num4     0
num5     0
num6     0
num7     0
num8     0
num9     0
num10    0
num11    0
num12    0
num13    0
num14    0
num15    0
num16    0
num17    0
num19    0
num20    0
num21    0
num23    0
der1     0
der2     0
der3     0
der4     0
der5     0
der6     0
der7     0
der8     0
der9     0
der10    0
der11    0
der12    0
der13    0
der14    0
der15    0
der16    0
der17    0
der18    0
der19    0
cat1     0
cat2     0
cat4     0
cat5     0
cat7     0
cat9     0
cat11    0
cat12    0
cat13    0
cat14    0
dtype: int64

In [105]:
## data_train now consists of 51 columns including target filed and free of any missing or null values.
print(data_train.head())

## data_test now consists of 50 columns excluding target field and free of any missing or null values.
data_test.head()

   num1  num2  num3  num4  num5  num6  num7  num8  num9  num10   ...    cat2  \
0     2     5     0     1     0     0     0     0     0      0   ...     1.0   
1     1     7     0     0     1     0     0     0     0      0   ...     0.0   
2     5     9     0     0     1     0     0     0     0      0   ...     1.0   
3     0     2     1     0     0     0     0     0     0      0   ...     0.0   
4     0     0     1     0     0     0     0     0     0      0   ...     1.0   

   cat4  cat5  cat7  cat9  cat11  cat12  cat13  cat14  target  
0  10.0   1.0     0     4      0    0.0      1     12       0  
1  11.0   1.0     0    11      1    2.0      1     19       0  
2   7.0   1.0     0    14      1    2.0      1     60       0  
3   7.0   1.0     0    11      1    3.0      1    104       0  
4  11.0   1.0     0    14      1    2.0      1     82       0  

[5 rows x 51 columns]


num1  num2  num3  num4  num5  num6  num7  num8  num9  num10  ...    cat1  \
0     0     8     0     1     0     0     0     0     0      0  ...     1.0   
1     4     5     0     0     0     1     0     0     0      0  ...     2.0   
2     5     3     0     0     0     1     0     0     0      0  ...     1.0   
3     0     6     1     0     0     0     0     0     0      0  ...     1.0   
4     5     7     0     0     0     1     0     0     0      0  ...     1.0   

   cat2  cat4  cat5  cat7  cat9  cat11  cat12  cat13  cat14  
0   1.0   7.0   1.0     0     1      1    2.0      1     65  
1   1.0   4.0   1.0     0    11      1    0.0      1    103  
2   0.0  11.0   1.0     0    14      1    2.0      1     29  
3   0.0   7.0   1.0     0     1      1    2.0      1     40  
4   0.0  11.0   1.0     0    11      1    2.0      1    101  

[5 rows x 50 columns]

Now we have a datasets with no missing values.
Our next step is to checking and making sure that all datas are in correct format (int , float , object).
Now lets understand which attributes are contributing or lets say affecting the target field.

The **Pearson Correlation** (corr() function)  measures the linear dependence between two variables X and Y.<br>
1: total positive linear correlation <br>
0: no linear correlation, the two variables most likely do not affect each other<br>
-1: total negative linear correlation.

It is difficult to get the correlation idea with so much attribute , so lets take the correlation idea by separating numerical , categorical and derived attributes.

In [107]:
## CORRELATION between numerical variables and target field:
data_train[['num1','num2','num3','num4','num5','num6','num7','num8','num9','num10','num11','num12','num13','num14','num15',
             'num16','num17','num19','num20','num21','num23','target' ]].corr()

num1      num2      num3      num4      num5      num6      num7  \
num1    1.000000  0.223337 -0.277461  0.005712  0.188540  0.162826  0.001846   
num2    0.223337  1.000000 -0.045234 -0.060255  0.014088  0.111223  0.006896   
num3   -0.277461 -0.045234  1.000000 -0.473988 -0.356854 -0.384352 -0.009021   
num4    0.005712 -0.060255 -0.473988  1.000000 -0.260436 -0.280505  0.010857   
num5    0.188540  0.014088 -0.356854 -0.260436  1.000000 -0.211185  0.006431   
num6    0.162826  0.111223 -0.384352 -0.280505 -0.211185  1.000000 -0.006994   
num7    0.001846  0.006896 -0.009021  0.010857  0.006431 -0.006994  1.000000   
num8   -0.016736 -0.003903 -0.005203  0.010150  0.009987 -0.014388  0.064624   
num9   -0.023377  0.001719 -0.017478  0.036612  0.012246 -0.030866  0.098625   
num10  -0.008760  0.003954 -0.008421  0.012200  0.007861 -0.010622  0.083977   
num11  -0.024942  0.002045 -0.018321  0.035604  0.015369 -0.031649  0.267418   
num12   0.122113  0.170438  0.016367 -0.120684  0.052761  0.064877 -0.022237   
num13   0.036913 -0.018345  0.028335 -0.072319  0.005786  0.040191 -0.024074   
num14   0.101256  0.027692 -0.037130  0.020843  0.036015 -0.011069  0.005320   
num15  -0.112216  0.001409 -0.011735  0.076506 -0.033159 -0.039693  0.027389   
num16   0.102138  0.003161 -0.020731 -0.030285 -0.136679  0.190358  0.004328   
num17   0.183925  0.048302 -0.183467  0.076783 -0.000213  0.144547  0.011646   
num19   0.008270 -0.059286  0.019702  0.067428 -0.082170 -0.022316  0.002586   
num20   0.161873  0.089084 -0.218241  0.208829  0.098200 -0.054004  0.015199   
num21   0.099759  0.048370 -0.272232  0.347781  0.049593 -0.096073  0.019656   
num23  -0.056636 -0.048033 -0.165985  0.196848  0.015961 -0.027879  0.001043   
target  0.018571  0.008283 -0.034026  0.034255  0.013135 -0.008254  0.001794   

            num8      num9     num10    ...        num13     num14     num15  \
num1   -0.016736 -0.023377 -0.008760    ...     0.036913  0.101256 -0.112216   
num2   -0.003903  0.001719  0.003954    ...    -0.018345  0.027692  0.001409   
num3   -0.005203 -0.017478 -0.008421    ...     0.028335 -0.037130 -0.011735   
num4    0.010150  0.036612  0.012200    ...    -0.072319  0.020843  0.076506   
num5    0.009987  0.012246  0.007861    ...     0.005786  0.036015 -0.033159   
num6   -0.014388 -0.030866 -0.010622    ...     0.040191 -0.011069 -0.039693   
num7    0.064624  0.098625  0.083977    ...    -0.024074  0.005320  0.027389   
num8    1.000000  0.250133  0.179093    ...    -0.054034  0.003609  0.068870   
num9    0.250133  1.000000  0.152086    ...    -0.099457  0.025179  0.109275   
num10   0.179093  0.152086  1.000000    ...    -0.040695  0.004111  0.050873   
num11   0.564804  0.890035  0.426928    ...    -0.106247  0.022042  0.121425   
num12  -0.061180 -0.116980 -0.040535    ...     0.312420 -0.021854 -0.451666   
num13  -0.054034 -0.099457 -0.040695    ...     1.000000 -0.518060 -0.594279   
num14   0.003609  0.025179  0.004111    ...    -0.518060  1.000000 -0.158009   
num15   0.068870  0.109275  0.050873    ...    -0.594279 -0.158009  1.000000   
num16   0.007110  0.013734  0.002841    ...    -0.056207  0.069967  0.020008   
num17   0.002501  0.007016  0.005431    ...     0.006311  0.033381 -0.027077   
num19   0.001214  0.004478  0.001273    ...     0.021156  0.008747 -0.026153   
num20   0.003492  0.018479  0.009936    ...    -0.037136  0.055922  0.000290   
num21   0.001102  0.022779  0.010881    ...    -0.038493  0.055689  0.001590   
num23  -0.008901 -0.001487 -0.003172    ...     0.007761  0.009940 -0.020107   
target  0.002012  0.007977  0.002448    ...    -0.027725  0.037056  0.004518   

           num16     num17     num19     num20     num21     num23    target  
num1    0.102138  0.183925  0.008270  0.161873  0.099759 -0.056636  0.018571  
num2    0.003161  0.048302 -0.059286  0.089084  0.048370 -0.048033  0.008283  
num3   -0.020731 -0.183467  0.019702 -0.218241 -0.272232 -0.165985 -0.034026  
num4   -0.0302

In [108]:
## Correlation between derived variables and target field:
data_train[['der1','der2','der3','der4','der5','der6','der7','der8','der9','der10','der11','der12','der13','der14',
             'der15','der16','der17','der18','der19','target']].corr()

der1      der2      der3      der4      der5      der6      der7  \
der1    1.000000 -0.002326  0.000526  0.002638  0.000571 -0.000229  0.000060   
der2   -0.002326  1.000000  0.002993  0.000121 -0.001636  0.000705  0.001764   
der3    0.000526  0.002993  1.000000 -0.000087 -0.001462  0.001011  0.002455   
der4    0.002638  0.000121 -0.000087  1.000000  0.000650 -0.001848  0.001518   
der5    0.000571 -0.001636 -0.001462  0.000650  1.000000 -0.001437 -0.002282   
der6   -0.000229  0.000705  0.001011 -0.001848 -0.001437  1.000000 -0.000131   
der7    0.000060  0.001764  0.002455  0.001518 -0.002282 -0.000131  1.000000   
der8    0.000322 -0.000110  0.000359 -0.001899 -0.001847 -0.000589 -0.002074   
der9   -0.001014  0.001213 -0.001247  0.000014  0.001123  0.001804  0.000805   
der10  -0.000312 -0.000489  0.000164 -0.000004 -0.000241 -0.001519 -0.000564   
der11  -0.001487 -0.000872 -0.000408  0.000250  0.000470 -0.000520 -0.000097   
der12  -0.002196 -0.000723  0.000174  0.000976 -0.000321  0.000568 -0.000312   
der13   0.000162  0.000761  0.001502 -0.000268  0.000272  0.000916 -0.000667   
der14   0.000681  0.000374 -0.000529  0.001182 -0.000926 -0.000825  0.001904   
der15   0.000546  0.000019  0.001376  0.000157  0.000123  0.000660  0.000166   
der16   0.000135  0.001831 -0.000181 -0.000840 -0.001012  0.000113 -0.001348   
der17  -0.000445 -0.000430 -0.000989  0.001294 -0.000930  0.000337 -0.000329   
der18   0.000502  0.001303 -0.001799  0.001165 -0.000427  0.001990 -0.000304   
der19  -0.001337  0.000331  0.000503  0.000901  0.000204  0.001187 -0.001193   
target  0.001814  0.001364  0.001921  0.000023  0.000756  0.000113 -0.000101   

            der8      der9     der10     der11     der12     der13     der14  \
der1    0.000322 -0.001014 -0.000312 -0.001487 -0.002196  0.000162  0.000681   
der2   -0.000110  0.001213 -0.000489 -0.000872 -0.000723  0.000761  0.000374   
der3    0.000359 -0.001247  0.000164 -0.000408  0.000174  0.001502 -0.000529   
der4   -0.001899  0.000014 -0.000004  0.000250  0.000976 -0.000268  0.001182   
der5   -0.001847  0.001123 -0.000241  0.000470 -0.000321  0.000272 -0.000926   
der6   -0.000589  0.001804 -0.001519 -0.000520  0.000568  0.000916 -0.000825   
der7   -0.002074  0.000805 -0.000564 -0.000097 -0.000312 -0.000667  0.001904   
der8    1.000000  0.000017 -0.002635  0.002067  0.000036 -0.000304 -0.000188   
der9    0.000017  1.000000 -0.000231 -0.000861 -0.001417 -0.000701  0.000019   
der10  -0.002635 -0.000231  1.000000  0.000415  0.002009  0.001358  0.001291   
der11   0.002067 -0.000861  0.000415  1.000000  0.001031  0.003667  0.000470   
der12   0.000036 -0.001417  0.002009  0.001031  1.000000 -0.000280  0.000964   
der13  -0.000304 -0.000701  0.001358  0.003667 -0.000280  1.000000 -0.002345   
der14  -0.000188  0.000019  0.001291  0.000470  0.000964 -0.002345  1.000000   
der15  -0.000312  0.000860 -0.000333  0.000817  0.001588  0.000831  0.001656   
der16   0.002127 -0.001284  0.000343  0.000909 -0.001034  0.002557  0.000254   
der17  -0.000577 -0.000804  0.000414 -0.001199  0.003580 -0.000690 -0.001326   
der18   0.000186 -0.001080 -0.001735 -0.001059  0.000162  0.000529 -0.001070   
der19   0.000798  0.000483 -0.001259 -0.001361  0.000952  0.001875 -0.000091   
target -0.000960  0.000723  0.001028  0.000384 -0.001195 -0.000430  0.001409   

           der15     der16     der17     der18     der19    target  
der1    0.000546  0.000135 -0.000445  0.000502 -0.001337  0.001814  
der2    0.000019  0.001831 -0.000430  0.001303  0.000331  0.001364  
der3    0.001376 -0.000181 -0.000989 -0.001799  0.000503  0.001921  
der4    0.000157 -0.000840  0.001294  0.001165  0.000901  0.000023  
der5    0.000123 -0.001012 -0.000930 -0.000427  0.000204  0.000756  
der6    0.000660  0.000113  0.000337  0.001990  0.001187  0.000113  
der7    0.000166 -0.001348 -0.000329 -0.000304 -0.001193 -0.000101  
der8   -0.000312  0.002127 -0.000577  0.000186  0.000798 -0.000960  
der9    0.000860 -0.001284 -0

In [109]:
## Correlation between categorical variables and target field:
data_train[['cat1','cat2','cat4','cat5','cat7','cat9','cat11','cat12','cat13','cat14','target']].corr()

cat1      cat2      cat4      cat5      cat7      cat9     cat11  \
cat1    1.000000  0.146343 -0.000045  0.045273 -0.071957 -0.017431  0.001457   
cat2    0.146343  1.000000  0.027965  0.003102 -0.045141  0.025305 -0.008250   
cat4   -0.000045  0.027965  1.000000 -0.152969  0.068641  0.053266 -0.059856   
cat5    0.045273  0.003102 -0.152969  1.000000 -0.273597 -0.253074  0.036120   
cat7   -0.071957 -0.045141  0.068641 -0.273597  1.000000  0.193997 -0.051065   
cat9   -0.017431  0.025305  0.053266 -0.253074  0.193997  1.000000 -0.014601   
cat11   0.001457 -0.008250 -0.059856  0.036120 -0.051065 -0.014601  1.000000   
cat12   0.007448 -0.042874  0.273495 -0.014153 -0.037382  0.003545 -0.017291   
cat13   0.011179  0.022780  0.006001 -0.019155  0.000728 -0.003222 -0.031213   
cat14   0.015213 -0.000467  0.010968 -0.000936  0.070465  0.025358  0.034401   
target  0.005421  0.009941  0.017628 -0.031532  0.032909  0.011579 -0.020300   

           cat12     cat13     cat14    target  
cat1    0.007448  0.011179  0.015213  0.005421  
cat2   -0.042874  0.022780 -0.000467  0.009941  
cat4    0.273495  0.006001  0.010968  0.017628  
cat5   -0.014153 -0.019155 -0.000936 -0.031532  
cat7   -0.037382  0.000728  0.070465  0.032909  
cat9    0.003545 -0.003222  0.025358  0.011579  
cat11  -0.017291 -0.031213  0.034401 -0.020300  
cat12   1.000000 -0.023784 -0.010725  0.006133  
cat13  -0.023784  1.000000  0.003478  0.000956  
cat14  -0.010725  0.003478  1.000000  0.006105  
target  0.006133  0.000956  0.006105  1.000000

 **P-value**: 
 The P-value is the probability value that the correlation between these two variables is statistically significant. Normally, we choose a significance level of 0.05, which means that we are 95% confident that the correlation between the variables is significant. 

By convention, when the 
- p-value is < 0.001 we say there is strong evidence that the correlation is significant,
- the p-value is < 0.05; there is moderate evidence that the correlation is significant,
- the p-value is < 0.1; there is weak evidence that the correlation is significant, and
- the p-value is > 0.1; there is no evidence that the correlation is significant.

#### Calculating the P-value and and Pearson Coefficient for all the columns field with target filed to check the statistical and linear significance of each variable with target field.

In [10]:
from scipy import stats
for i in  range(0,51,1):
     pearson_coef, p_value = stats.pearsonr(data_train[data_train.columns[i]],data_train['target'])
     print("The Pearson Correlation Coefficient of ",data_train.columns[i],  pearson_coef, " with a P-value of P =", p_value) 

The Pearson Correlation Coefficient of  num1 0.01857143788602922  with a P-value of P = 1.2565539539334194e-46
The Pearson Correlation Coefficient of  num2 0.00828314230031639  with a P-value of P = 1.6079133361820057e-10
The Pearson Correlation Coefficient of  num3 -0.03402621646622292  with a P-value of P = 3.5935756239788726e-152
The Pearson Correlation Coefficient of  num4 0.03425511115747301  with a P-value of P = 3.369288318085845e-154
The Pearson Correlation Coefficient of  num5 0.01313468598373151  with a P-value of P = 3.6511395384758e-24
The Pearson Correlation Coefficient of  num6 -0.008254381095529822  with a P-value of P = 1.8589510311803223e-10
The Pearson Correlation Coefficient of  num7 0.001794067589473939  with a P-value of P = 0.16604116009214964
The Pearson Correlation Coefficient of  num8 0.0020117232635613607  with a P-value of P = 0.12040652687507707
The Pearson Correlation Coefficient of  num9 0.007977067632197967  with a P-value of P = 7.345390824060392e-10
The

Almost all the variables are showing weak correlation (i.e linear realtionship) . Now we will sort out the variable having **strong Statistical significance** i.e. P-value < 0.001  with target variable:


#### Following are the Numerical Variable with strong Statistical Significance (P-Value < 0.001): 
num1    P-value of P = 1.2565539539334194e-46  <br>
num2    P-value of P = 1.6079133361820057e-10  <br>
num3    P-value of P = 3.5935756239788726e-152<br>
num4    P-value of P = 3.369288318085845e-154<br>
num5    P-value of P = 3.6511395384758e-24<br>
num6    P-value of P = 1.8589510311803223e-10<br>
num9    P-value of P = 7.345390824060392e-10<br>
num11   P-value of P = 5.437383527994713e-09<br>
num12   P-value of P = 3.0661197812294255e-62<br>
num13   P-value of P = 1.132853956644825e-101  <br>
num14   P-value of P = 4.07628834505048e-180<br>
num15   P-value of P = 0.00048715410649929707<br>
num16   P-value of P = 3.4275683065295814e-70<br>
num17   P-value of P = 5.35926716001635e-159<br>
num20   P-value of P = 1.8445393191972405e-197<br>
num23   P-value of P = 1.7236484191707134e-101<br>
#### All the Derived Variable are showing no evidence of Statistical Significance (P-Value > 0.1): <br>
#### Following are the Categorical Variable with strong Statistical Significance (P-Value < 0.001): 
cat1    P-value of P = 2.847843593932867e-05 <br>
cat2    P-value of P = 1.6563120922775573e-14 <br>
cat4    P-value of P = 3.476136449955979e-42 <br>
cat5    P-value of P = 5.888546104559579e-131 <br>
cat7    P-value of P = 1.8200283213936985e-142 <br>
cat9    P-value of P = 3.919166362342092e-19 <br>
cat11   P-value of P = 2.3044521962053904e-55 <br>
cat12   P-value of P = 2.1959849360013953e-06 <br>
cat14   P-value of P = 2.4416539468153042e-06 <br>

We will give more emphasis on these variables while predicting the target field:

We are now going to apply various **Machine Learning** Algorithm on the **data_train** and will se how closely the target field is getting predicted :

### Approach 1: Taking all the data fields and applying Machine Learning algorithm to it: 

In [8]:
## Assigning X_train , Y_train and checking whether the data are in proper format or not i.e no of columns of X_train 
#   and data_test

X_train=data_train.drop('target',axis=1)
Y_train=data_train['target']
#X_test=data_test.drop('id',axis=1).copy()
X_train.shape,Y_train.shape,data_test.shape

((596000, 50), (596000,), (892816, 50))

In [12]:
# Logistic Regression
from sklearn.linear_model import LogisticRegression

logreg = LogisticRegression()
logreg.fit(X_train, Y_train)
Y_predlog = logreg.predict(data_test)
acc_log = round(logreg.score(X_train, Y_train) * 100, 2)
acc_log

96.36

In [13]:
## Gaussian Naive Bayes
from sklearn.naive_bayes import GaussianNB

gaussian = GaussianNB()
gaussian.fit(X_train, Y_train)
Y_predGNB = gaussian.predict(data_test)
acc_gaussian = round(gaussian.score(X_train, Y_train) * 100, 2)
acc_gaussian

91.43

In [14]:
# Perceptron
from sklearn.linear_model import Perceptron

perceptron = Perceptron()
perceptron.fit(X_train, Y_train)
Y_predPCT = perceptron.predict(data_test)
acc_perceptron = round(perceptron.score(X_train, Y_train) * 100, 2)
acc_perceptron

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.perceptron.Perceptron'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


93.22

In [18]:
## Linear SVC
from sklearn.svm import  LinearSVC

linear_svc = LinearSVC()
linear_svc.fit(X_train, Y_train)
Y_predlSVC = linear_svc.predict(data_test)
acc_linear_svc = round(linear_svc.score(X_train, Y_train) * 100, 2)
acc_linear_svc

96.36

In [15]:
# Stochastic Gradient Descent
from sklearn.linear_model import SGDClassifier

sgd = SGDClassifier()
sgd.fit(X_train, Y_train)
Y_predSGD = sgd.predict(data_test)
acc_sgd = round(sgd.score(X_train, Y_train) * 100, 2)
acc_sgd

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


44.8

In [9]:
# Decision Tree
from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier()
decision_tree.fit(X_train, Y_train)
Y_predDTree = decision_tree.predict(data_test)
acc_decision_tree = round(decision_tree.score(X_train, Y_train) * 100, 2)
acc_decision_tree

100.0

In [17]:
# Random Forest
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(n_estimators=100)
random_forest.fit(X_train, Y_train)
Y_predRfor = random_forest.predict(data_test)
random_forest.score(X_train, Y_train)
acc_random_forest = round(random_forest.score(X_train, Y_train) * 100, 2)
acc_random_forest

99.99

In [21]:
#Model Evaluation
models = pd.DataFrame({
    'Model': ['Logistic Regression','Naive Bayes', 'Perceptron',
              'Linear SVC', 'Stochastic Gradient Decent', 'Decision Tree',
              'Random Forest'],
    'Score': [acc_log, acc_gaussian, acc_perceptron, 
              acc_linear_svc, acc_sgd, acc_decision_tree,
              acc_random_forest]})
models.sort_values(by='Score', ascending=False)

Model   Score
5               Decision Tree  100.00
6               Random Forest   99.99
0         Logistic Regression   96.36
3                  Linear SVC   96.36
2                  Perceptron   93.22
1                 Naive Bayes   91.43
4  Stochastic Gradient Decent   44.80

In [10]:
submission = pd.DataFrame({
        "id": data_test_id,
        "target": Y_predDTree
    })
submission.head()

id  target
0   0       0
1   1       0
2   2       0
3   3       0
4   4       0

In [30]:
submission.to_csv('E:/softwares/Ubuntu/dataset/quartic/ds_data/submission_aprr1.csv' , index=False)

In [31]:
### Checking the value filed in the target field of submission , it shoud be either 1 or 0
print(submission['target'].nunique())
submission['target'].unique()

2


array([0, 1], dtype=int64)

In [32]:
submission.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 892816 entries, 0 to 892815
Data columns (total 2 columns):
id        892816 non-null int64
target    892816 non-null int64
dtypes: int64(2)
memory usage: 13.6 MB


### Approach 2: Taking only the data fields with high statistical significance  and applying Machine Learning algorithm to it: 

In [11]:
data_train1 = data_train[['num1','num2','num3','num4','num5','num6','num9','num11','num12','num13','num14','num15','num16',
                           'num17','num20','num23','cat1','cat2','cat4','cat5','cat7','cat9','cat11','cat12','cat14',
                            'target']]
data_test1=data_test[['num1','num2','num3','num4','num5','num6','num9','num11','num12','num13','num14','num15','num16',
                           'num17','num20','num23','cat1','cat2','cat4','cat5','cat7','cat9','cat11','cat12','cat14']]
print('Done')

Done


In [12]:
 ## Assigning X_train , Y_train and checking whether the data are in proper format or not i.e no of columns of X_train 
#   and data_test

X_train1=data_train1.drop('target',axis=1)
Y_train1=data_train1['target']
#X_test=data_test.drop('id',axis=1).copy()
X_train1.shape,Y_train1.shape,data_test1.shape

((596000, 25), (596000,), (892816, 25))

In [35]:
# Logistic Regression
from sklearn.linear_model import LogisticRegression

logreg = LogisticRegression()
logreg.fit(X_train1, Y_train1)
Y_predlog = logreg.predict(data_test1)
acc_log = round(logreg.score(X_train1, Y_train1) * 100, 2)
acc_log

96.36

In [36]:
## Gaussian Naive Bayes
from sklearn.naive_bayes import GaussianNB

gaussian = GaussianNB()
gaussian.fit(X_train1, Y_train1)
Y_predGNB = gaussian.predict(data_test1)
acc_gaussian = round(gaussian.score(X_train1, Y_train1) * 100, 2)
acc_gaussian

91.9

In [37]:
# Perceptron
from sklearn.linear_model import Perceptron

perceptron = Perceptron()
perceptron.fit(X_train1, Y_train1)
Y_predPCT = perceptron.predict(data_test1)
acc_perceptron = round(perceptron.score(X_train1, Y_train1) * 100, 2)
acc_perceptron

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.perceptron.Perceptron'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


95.28

In [38]:
## Linear SVC
from sklearn.svm import  LinearSVC

linear_svc = LinearSVC()
linear_svc.fit(X_train1, Y_train1)
Y_predlSVC = linear_svc.predict(data_test1)
acc_linear_svc = round(linear_svc.score(X_train1, Y_train1) * 100, 2)
acc_linear_svc

96.36

In [39]:
# Stochastic Gradient Descent
from sklearn.linear_model import SGDClassifier

sgd = SGDClassifier()
sgd.fit(X_train1, Y_train1)
Y_predSGD = sgd.predict(data_test1)
acc_sgd = round(sgd.score(X_train1, Y_train1) * 100, 2)
acc_sgd

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


96.33

In [15]:
# Decision Tree
from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier()
decision_tree.fit(X_train1, Y_train1)
Y_predDTree = decision_tree.predict(data_test1)
acc_decision_tree = round(decision_tree.score(X_train1, Y_train1) * 100, 2)
acc_decision_tree

99.99

In [41]:
# Random Forest
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(n_estimators=100)
random_forest.fit(X_train1, Y_train1)
Y_predRfor = random_forest.predict(data_test1)
random_forest.score(X_train, Y_train)
acc_random_forest = round(random_forest.score(X_train1, Y_train1) * 100, 2)
acc_random_forest

99.99

In [42]:
#Model Evaluation
models = pd.DataFrame({
    'Model': ['Logistic Regression','Naive Bayes', 'Perceptron',
              'Linear SVC', 'Stochastic Gradient Decent', 'Decision Tree',
              'Random Forest'],
    'Score': [acc_log, acc_gaussian, acc_perceptron, 
              acc_linear_svc, acc_sgd, acc_decision_tree,
              acc_random_forest]})
models.sort_values(by='Score', ascending=False)

Model  Score
5               Decision Tree  99.99
6               Random Forest  99.99
0         Logistic Regression  96.36
3                  Linear SVC  96.36
4  Stochastic Gradient Decent  96.33
2                  Perceptron  95.28
1                 Naive Bayes  91.90

In [16]:
submission1 = pd.DataFrame({
        "id": data_test_id,
        "target": Y_predDTree
    })
submission1.head()

id  target
0   0       0
1   1       0
2   2       0
3   3       0
4   4       1

In [44]:
submission1.to_csv('E:/softwares/Ubuntu/dataset/quartic/ds_data/submission_aprr2.csv' , index=False)

In [46]:
### Checking the value field in the target field of submission , it shoud be either 1 or 0
print(submission1['target'].nunique())
submission1['target'].unique()

2


array([0, 1], dtype=int64)

In [47]:
submission1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 892816 entries, 0 to 892815
Data columns (total 2 columns):
id        892816 non-null int64
target    892816 non-null int64
dtypes: int64(2)
memory usage: 13.6 MB


### Train-Test Split

To understand which approach is predicting precise value of of the data_test. We will split the data_train into test and train. 
Doing so we will have actual value and predicted value and we will see how close we are in terms of accuracy.


#### Approach 1 : Taking all the field and applying Decision Tree on data_train :

In [23]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(X_train, Y_train, test_size=0.45, random_state=0)
print("done")

# We will perform decision tree ML on x_test data. 

decision_tree0 = DecisionTreeClassifier()
decision_tree0.fit(x_train, y_train)
Y_predDTree0 = decision_tree0.predict(x_test)
acc_decision_tree = round(decision_tree0.score(x_train, y_train) * 100, 2)
acc_decision_tree

done


100.0

Now we have the actual value (y_test) and predicted value (Y_predDTree0) on the test data i.e x_test. We will compare this and understand how close our prediction is :

In [45]:
Y_predDTree0=pd.DataFrame({
        "target pred": Y_predDTree0})
Y_predDTree0.head()

target pred
0            0
1            0
2            0
3            0
4            0

In [44]:
#y_test = pd.DataFrame({
#        "target1": y_test
#    })
y_test.head()
#y_test=y_test.reset_index()
#y_test.head()
del y_test['index']
y_test.head()

target
0       0
1       0
2       0
3       0
4       0

We are going to compare **Y_predDTree0["target pred"]** and **y_test["target"]** . If any mismatch then we will take that into consideration in form of count. The less number of mismatch suggests better model:

In [65]:
#268200 is the total number of records in Y_predDTree0 and y_test

count=0

for i in range(0,268200,1):
    if(Y_predDTree0.loc[i,'target pred'] != y_test.loc[i,'target']):
        count=count + 1
        
print(count)

22321


Thus we can conclude that for 22321 out of 268200 number of records our prediction is wrong. Thus our model accuracy using all the fields is  approx 91.6% , it means  for approx 8.4% our prediction is going to be wrong. We can say that 91.6% of accuracy is a fair result.

#### Approach 2: Taking only the fields with high statistical significance i.e data_train1 and applying ML to it  : 

In [22]:
#from sklearn.model_selection import train_test_split

x_train1, x_test1, y_train1, y_test1 = train_test_split(X_train1, Y_train1, test_size=0.45, random_state=0)
print("done")

# We will perform decision tree ML on x_test1 data. 

decision_tree1 = DecisionTreeClassifier()
decision_tree1.fit(x_train1, y_train1)
Y_predDTree1 = decision_tree.predict(x_test1)
acc_decision_tree = round(decision_tree1.score(x_train1, y_train1) * 100, 2)
acc_decision_tree

done


99.99

In [46]:
Y_predDTree1=pd.DataFrame({
        "target pred": Y_predDTree1})
Y_predDTree1.head()

target pred
0            0
1            0
2            0
3            0
4            0

In [50]:
#y_test1 = pd.DataFrame({
#        "target": y_test1
#    })
#print(y_test1.head())
#y_test1=y_test1.reset_index()
del y_test1['level_0']
y_test1.head()

target
0       0
1       0
2       0
3       0
4       0

In [69]:
#268200 is the total number of records in Y_predDTree1 and y_test1

count1=0

for i in range(0,268200,1):
    if(Y_predDTree1.loc[i,'target pred'] != y_test1.loc[i,'target']):
        count1=count1 + 1
        
print(count1)

21


Her we can conclude that for only 21 out of 268200 number of records our prediction is wrong. Thus our model accuracy **using only fields with high statistical significace is nearly ~100%**.

#### Thus we can conclude that the model is giving great accuracy with the high statistical significance field only.


###  Hence submission1 is the final submission for this dataset which is Approach 2. 